File: php/test/SecurityEntityTest.php
<?php
declare(strict_types=1);

// Security entity test

require_once __DIR__ . '/../dymoapiintroduction_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class SecurityEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = DymoApiIntroductionSDK::test(null, null);
        $ent = $testsdk->Security(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = security_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "security." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $security_ref01_ent = $client->Security(null);
        $security_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.security"), "security_ref01"));

        $security_ref01_data_result = $security_ref01_ent->create($security_ref01_data, null);
        $security_ref01_data = Helpers::to_map(is_object($security_ref01_data_result) && method_exists($security_ref01_data_result, 'data_get') ? $security_ref01_data_result->data_get() : $security_ref01_data_result);
        $this->assertNotNull($security_ref01_data);

    }
}

function security_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/security/SecurityTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = DymoApiIntroductionSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["security01", "security02", "security03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID" => $idmap,
        "DYMO_API_INTRODUCTION_TEST_LIVE" => "FALSE",
        "DYMO_API_INTRODUCTION_TEST_EXPLAIN" => "FALSE",
        "DYMO_API_INTRODUCTION_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["DYMO_API_INTRODUCTION_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["DYMO_API_INTRODUCTION_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new DymoApiIntroductionSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["DYMO_API_INTRODUCTION_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["DYMO_API_INTRODUCTION_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/security_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/dymo-api-introduction-sdk/go"
	"github.com/voxgig-sdk/dymo-api-introduction-sdk/go/core"

	vs "github.com/voxgig-sdk/dymo-api-introduction-sdk/go/utility/struct"
)

func TestSecurityEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Security(nil)
		if ent == nil {
			t.Fatal("expected non-nil SecurityEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := securityBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "security." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		securityRef01Ent := client.Security(nil)
		securityRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "security"}, setup.data), "security_ref01"))

		securityRef01DataResult, err := securityRef01Ent.Create(securityRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		securityRef01Data = core.ToMapAny(entityData(securityRef01DataResult))
		if securityRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

	})
}

func securityBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "security", "SecurityTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read security test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse security test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"security01", "security02", "security03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID": idmap,
		"DYMO_API_INTRODUCTION_TEST_LIVE":      "FALSE",
		"DYMO_API_INTRODUCTION_TEST_EXPLAIN":   "FALSE",
		"DYMO_API_INTRODUCTION_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["DYMO_API_INTRODUCTION_TEST_SECURITY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["DYMO_API_INTRODUCTION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["DYMO_API_INTRODUCTION_APIKEY"],
			},
			extra,
		})
		client = sdk.NewDymoApiIntroductionSDK(core.ToMapAny(mergedOpts))
	}

	live := env["DYMO_API_INTRODUCTION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["DYMO_API_INTRODUCTION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
